### PR TITLE
Add `Verified` filter to search filters

### PR DESCRIPTION
**Pull request by oisincoveney**
Epic: https://github.com/metabase/metabase/issues/27982

### Description

Adds `Verified` filter to search modal for Pro/EE instances.

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Start an enterprise instance with `yarn dev-ee` - ensure that you have an enterprise token (instructions are [here](https://www.notion.so/metabase/Cypress-How-To-4d1c70bb0ce94a60b3688fd7acba8471?pvs=4))
2. Verify that the `Verified` filter shows up on the search modal, and searching with verified filter only displays verified items.

### Checklist

- [X] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Review by ranquild — CHANGES_REQUESTED** [review]
We need to move as much code as possible to the EE folder for premium features. Not sure what the best way to do it here, but I would try moving `VerifiedFilter` and exposing it via the plugin, e.g. not

```
[SearchFilterKeys.Verified]: VerifiedFilter,
```

but
```
[SearchFilterKeys.Verified]: PLUGIN_CONTENT_VERIFICATION.VerifiedFilter,
```

btw it won't work if you initialize it as a global constant. You would need to move it to a function.

**Comment by oisincoveney** [comment]
> We need to move as much code as possible to the EE folder for premium features. Not sure what the best way to do it here, but I would try moving `VerifiedFilter` and exposing it via the plugin, e.g. not
> 
> ```
> [SearchFilterKeys.Verified]: VerifiedFilter,
> ```
> 
> but
> 
> ```
> [SearchFilterKeys.Verified]: PLUGIN_CONTENT_VERIFICATION.VerifiedFilter,
> ```
> 
> btw it won't work if you initialize it as a global constant. You would need to move it to a function.

@ranquild I was actually able to follow your advice exactly 😄 let me know if there's anything else I should move over to the enterprise side, though not sure what else could go there

**Review by ranquild — COMMENTED** [review]
@oisincoveney looks good! Please update the tests using this https://www.notion.so/metabase/Testing-paid-features-49e03e1b2a3b41ffb4ebdc7616a93d90. There should be 3 test suites - OSS (no token), EE - a token but without features, and EE - a token with just the feature flag we need

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx [hunk 1]
```diff
@@ -29,16 +29,30 @@ export const SearchFilterModal = ({
...
+  );
+
+  const isValidFilterValue = useCallback(
+    (
+      key: FilterTypeKeys,
+      val: SearchFilterPropTypes[FilterTypeKeys],
+    ): boolean =>
+      !!val &&
+      (!Array.isArray(val) || val.length > 0) &&
+      !!filterMap[key as FilterTypeKeys],
```
Comment (oisincoveney) [inline frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:47]: This will prevent any possibility of `verified: true` appearing in the filter object if the user access the URL `/search?verified=true` in a non-premium instance

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx [hunk 2]
```diff
@@ -83,14 +98,14 @@ export const SearchFilterModal = ({
         </Flex>
         {availableFilters.map(key => {
           const Filter = filterMap[key];
-          return (
+          return Filter ? (
```
Comment (oisincoveney) [inline frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:101]: I'm going to separate this logic into a different component in the next filter PR and make this logic look a little cleaner

File: e2e/test/scenarios/onboarding/search/search.cy.spec.js [hunk 3]
```diff
@@ -132,9 +137,10 @@ describe("scenarios > search", () => {
     });
   });
 
-  describe("applying search filters", () => {
+  describeEE("applying search filters", () => {
```
Comment (WiNloSt) [inline e2e/test/scenarios/onboarding/search/search.cy.spec.js:140]: I don't think you should wrap everything inside `describeEE`. If you look at the current file on `master` https://github.com/metabase/metabase/blob/0f65cde6e714475995f126a0613730bc3694f92d/e2e/test/scenarios/onboarding/search/search.cy.spec.js

There are 11 `it()`s (1 of which is for a snowplow test)

After changing this to `describeEE`, only 4 tests are run on the opensource edition.
image

When comparing this to the run on master which has 17 tests (not counting the snowplow one). This change reduce the coverage on the open source edition significantly.
image

I think you should extract the enterprise test to a new `describeEE` block and leave the rest of the tests the same.

File: e2e/test/scenarios/onboarding/search/search.cy.spec.js [hunk 4]
```diff
@@ -162,6 +168,27 @@ describe("scenarios > search", () => {
...
+        cy.visit("/");
+        cy.findByTestId("search-bar-filter-button").click();
+        cy.findByTestId("verified-search-filter").should("exist");
+      });
+      it("should not render if Starter instance", () => {
+        setTokenFeatures("none");
+        cy.visit("/");
+        cy.findByTestId("search-bar-filter-button").click();
+        cy.findByTestId("verified-search-filter").should("not.exist");
+      });
```
Comment (WiNloSt) [inline e2e/test/scenarios/onboarding/search/search.cy.spec.js:191]: This test should be extracted out of `describeEE` otherwise, it won't run on open source edition.

File: e2e/test/scenarios/onboarding/search/search.cy.spec.js [hunk 5]
```diff
@@ -244,52 +271,143 @@ describe("scenarios > search", () => {
...
-            });
+          getSearchBar().clear().type("Orders{enter}");
+          cy.wait("@search");
+
+          cy.url().should("include", `verified=true`);
+
+          const verifiedItem = cy.findByTestId("search-result-item");
+          verifiedItem.within(() => {
+            cy.findByLabelText("verified icon").should("exist");
+          });
```
Comment (WiNloSt) [inline e2e/test/scenarios/onboarding/search/search.cy.spec.js:322]: If there is only one item inside `within` block we could make the syntax more concise by chaining the function call instead

```suggestion
          verifiedItem.findByLabelText("verified icon").should("exist");
```

File: e2e/test/scenarios/onboarding/search/search.cy.spec.js [hunk 6]
```diff
@@ -242,52 +269,143 @@ describe("scenarios > search", () => {
...
+            .each($el => {
+              if ($el.find('[aria-label="verified icon"]').length) {
+                verifiedElementCount++;
+              } else {
+                unverifiedElementCount++;
+              }
+            })
+            .then(() => {
+              expect(verifiedElementCount).to.eq(1);
+              expect(unverifiedElementCount).to.be.gt(0);
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply e2e/test/scenarios/onboarding/search/search.cy.spec.js:352]: I don't see we count verified/unverified items for the test `"should filter only for `Verified` assets"`. It feels like that test only checks if the verified item is found in the search result, but there's no apparent assertion to assert that there are **no** non-verified items in the search result.

I kind of expect the test body to be similar between these 2 tests since they're almost the same but just asserting different number of non-verified items.

File: enterprise/frontend/src/metabase-enterprise/content_verification/components/VerifiedSearchFilter/VerifiedSearchFilter.styled.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import styled from "@emotion/styled";
+import Button from "metabase/core/components/Button";
+import { color, lighten } from "metabase/lib/colors";
+
+export const VerifiedFilterButton = styled(Button)<{ isSelected: boolean }>`
+  ${({ isSelected }) =>
+    isSelected
+      ? `
```
Comment (WiNloSt) [inline enterprise/frontend/src/metabase-enterprise/content_verification/components/VerifiedSearchFilter/VerifiedSearchFilter.styled.tsx:8]: nit: using `css` from `@emotion/react` would make DX viewing the CSS rules better for supported IDEs. For example, these screenshots are from my setup (VS Code)

#### This PR
image

#### With `css` it's formatted like CSS
image

File: enterprise/frontend/src/metabase-enterprise/content_verification/components/VerifiedSearchFilter/VerifiedSearchFilter.tsx [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        data-is-selected={value}
+        isSelected={!!value}
+        onClick={() => onChange(true)}
+      >
+        Only verified items
+      </VerifiedFilterButton>
+      <VerifiedFilterButton
+        data-is-selected={value}
+        isSelected={!value}
+        onClick={() => onChange(undefined)}
```
Comment (WiNloSt) [inline enterprise/frontend/src/metabase-enterprise/content_verification/components/VerifiedSearchFilter/VerifiedSearchFilter.tsx:24]: `onChange(undefined)` looks precarious when seeing another usage is `onChange(true)`

**Review by WiNloSt — COMMENTED** [review]
Part 1 review. Will continue afterwards

File: frontend/src/metabase/nav/components/SearchBar.tsx [hunk 9]
```diff
@@ -16,7 +16,7 @@ import { zoomInRow } from "metabase/query_builder/actions";
 
 import { getSetting } from "metabase/selectors/settings";
 import RecentsList from "metabase/nav/components/RecentsList";
-import { SearchFilterModal } from "metabase/search/components/SearchFilterModal/SearchFilterModal";
+import { SearchFilterModal } from "metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal";
```
Comment (WiNloSt) [inline frontend/src/metabase/nav/components/SearchBar.tsx:19]: nit: I think you should do something about this `SearchFilterModal` repeating 3 times in the path. It's starting to get confusing.

File: frontend/src/metabase/search/types.ts [hunk 10]
```diff
@@ -16,9 +16,11 @@ export interface WrappedResult extends SearchResult {
 }
 
 export type TypeFilterProps = SearchModelType[];
+export type VerifiedFilterProps = boolean | undefined;
```
Comment (WiNloSt) [inline frontend/src/metabase/search/types.ts:19]: Can we get away with not using `undefined` for booleans? Using `undefined` suggests that we treat `undefined` differently from `true` and `false` and that could complicate a really simple boolean checks in many cases.

File: frontend/src/metabase/plugins/index.ts [hunk 11]
```diff
@@ -287,3 +287,7 @@ export const PLUGIN_MODEL_PERSISTENCE = {
 export const PLUGIN_EMBEDDING = {
   isEnabled: () => false,
 };
+
+export const PLUGIN_CONTENT_VERIFICATION = {
+  VerifiedFilter: null as any,
```
Comment (WiNloSt) [inline frontend/src/metabase/plugins/index.ts:292]: can we use something that's not `any`? Judging from `enterprise/frontend/src/metabase-enterprise/content_verification/components/VerifiedSearchFilter/VerifiedSearchFilter.tsx` I think this could be type as a React component if I'm not mistaken.

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal.unit.spec.tsx
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal.unit.spec.tsx:1]: I believed this is moved to `frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/tests/*` instead?

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx [hunk 12]
```diff
@@ -32,11 +29,30 @@ export const SearchFilterModal = ({
 }) => {
   const [output, setOutput] = useState<SearchFilters>(value);
 
+  const filterMap: Record<FilterTypeKeys, SearchFilterComponent> = useMemo(
+    () => ({
+      [SearchFilterKeys.Type]: TypeFilter,
+      [SearchFilterKeys.Verified]: PLUGIN_CONTENT_VERIFICATION.VerifiedFilter,
+    }),
+    [],
+  );
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:38]: Looks like it doesn't rely on any prop here, should we move it outside the component as a constant? That way, we don't need to memoize it either.

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx (hunk 1 above)
Comment (WiNloSt) [reply frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:47]: I think overall, this function is too hard to follow, can you simplify this more? You don't need to just cramp everything into a single expression. Using a longer-body function with lots of return is okay as long as it's readable.

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx [hunk 13]
```diff
@@ -64,11 +84,10 @@ export const SearchFilterModal = ({
...
 
-  // we can use this field to control which filters are available
-  // - we can enable the 'verified' filter here
-  const availableFilters: FilterTypeKeys[] = useMemo(() => {
-    return [SearchFilterKeys.Type];
-  }, []);
+  const availableFilters: FilterTypeKeys[] = useMemo(
+    () => [SearchFilterKeys.Type, SearchFilterKeys.Verified],
+    [],
+  );
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:90]: Should we lift this up outside the component as a constant as it doesn't rely on any props?

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx [hunk 14]
```diff
@@ -79,14 +98,14 @@ export const SearchFilterModal = ({
...
-          return (
+          return Filter ? (
             <Filter
               key={key}
               data-testid={`${key}-search-filter`}
               value={output[key]}
               onChange={value => onOutputChange(key, value)}
             />
-          );
+          ) : null;
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:108]: tbh based on how you control `availableFilters` and `filterMap` yourself here. I'm not sure why we need to check if this is null. It feels to me `Filter` will never be `undefined`

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/tests/setup.tsx [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+const TestSearchFilterModal = ({
+  initialFilters = {},
+  onChangeFilters,
+}: {
+  initialFilters?: SearchFilters;
+  onChangeFilters: jest.Mock;
+}) => {
+  const [filters, setFilters] = useState(initialFilters);
+
+  onChangeFilters.mockImplementation(newFilters => setFilters(newFilters));
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/tests/setup.tsx:53]: can't we simply wrap this in another function rather than mocking the implementation? This looks too complex.

e.g.
```suggestion
  function handleChangeFilters(newFilters) {
    setFilters(newFilters)
    onChangeFilter(newFilters)
  }
```

**Review by WiNloSt — COMMENTED** [review]
I've finished the review. Overall, I feels like there are places that could be cleaned up and simplified further.

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx (hunk 13 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:90]: There's some cleanup that I'm working on with the `created_by` and `created_at`, I'm going to put that as a separate PR so I could separate the logic of that from the general `Verified` filter stuff.

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal.unit.spec.tsx
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal.unit.spec.tsx:1]: Yeah the tests in this file were split between the files in that `tests` folder

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/tests/setup.tsx (hunk 15 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/tests/setup.tsx:53]: Yeah, I was back and forth on this one but my thinking was that I wanted to use as much jest as possible. I'll change it back, I have no real preferences here

File: e2e/test/scenarios/onboarding/search/search.cy.spec.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (oisincoveney) [reply e2e/test/scenarios/onboarding/search/search.cy.spec.js:352]: I see what you mean - the test for `should filter only for "Verified" assets` checks that _all_ elements within the search results have a `Verified` checkmark instead of counting how many search results we have. If there's at least one unverified element then it'll fail.

With the other test, we need to make sure that there's at least one of each result. So, we have to count the results to have some confidence that we're not just returning verified results when we know that we want to have "unverified" results too

File: e2e/test/scenarios/onboarding/search/search.cy.spec.js (hunk 3 above)
Comment (oisincoveney) [reply e2e/test/scenarios/onboarding/search/search.cy.spec.js:140]: Cool, I moved the tests around to test per-filter, and make sure that the `Verified` filter is the only one that relies on `describeEE` since that's the only one that is conditional on that

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx (hunk 12 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:38]: I'll clean this up in a separate PR (https://github.com/metabase/metabase/pull/33120#discussion_r1295610973)

File: frontend/src/metabase/search/types.ts (hunk 10 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/types.ts:19]: This is one that I'm grappling with - I didn't want to put `false` for two reasons:

* The backend expects `verified` to either be `true` or just not be there at all, and my thinking was that the frontend code should try and reflect the expected backend payload. 
* I think putting `false` here (and in the other comment with `onChange(undefined)`) is confusing. If we were to say `verified=false`, I think someone would come in and say "we're trying to search for unverified items" while the true behavior is that we're trying to get _all_ items.

What do you think? I personally think it's more confusing to put `false` instead of `undefined` because of the API, but I'd appreciate your perspective

File: frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx (hunk 14 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchFilterModal/SearchFilterModal/SearchFilterModal.tsx:108]: This will be resolved in the separate PR that I keep mentioning 😄

**Comment by oisincoveney** [comment]
@WiNloSt I left a few comments for the `undefined` vs `false` - let me know what you think on that one

File: frontend/src/metabase/search/types.ts (hunk 10 above)
Comment (WiNloSt) [reply frontend/src/metabase/search/types.ts:19]: sound reasonable to keep that in sync with BE. In that case, the type should be `true | undefined` instead. Since `boolean | undefined` implies 3 states are valid
1. `true`
2. `false`
3. `undefined`

Comment (oisincoveney) [reply frontend/src/metabase/search/types.ts:19]: Ah this is so much better. I'll fix that now.

File: e2e/test/scenarios/onboarding/search/search.cy.spec.js [hunk 16]
```diff
@@ -299,114 +238,199 @@ describe("scenarios > search", () => {
...
+        setTokenFeatures("all");
+        cy.createModerationReview({
+          status: "verified",
+          moderated_item_type: "card",
+          moderated_item_id: ORDERS_COUNT_QUESTION_ID,
         });
+      });
 
-        it("should show verified and unverified assets when `All items` is selected", () => {
+      describe("rendering on tiered instances", () => {
```
Comment (WiNloSt) [inline e2e/test/scenarios/onboarding/search/search.cy.spec.js:275]: Please see. https://metaboat.slack.com/archives/C057T1QTB3L/p1692185488242129

**Review by WiNloSt — APPROVED** [review]
Looks good to me now. Thanks for addressing the feedback 🎉

**Review by ranquild — APPROVED** [review]
Approving but with the recent updates this we'll likely get re-worked
